xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,153 @@
-     *            \- 14 -> Exponent given by next byte
-     *             - 15 -> Exponent given by next 2 bytes
+    if (end - *p < 8) {
+	throw Xapian::SerialisationError(
+	    "Bad encoded double: insufficient data");
+    }
+    double result;
+# ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(double));
+    memcpy(&temp, *p, sizeof(double));
+    temp = do_bswap(temp);
+    memcpy(&result, &temp, sizeof(double));
+# else
+    memcpy(&result, *p, sizeof(double));
+# endif
+    *p += 8;
+    return result;
+}
+
+#else
+
+string serialise_double(double v){
+    /*
+     * First bit(msb) -> sign (1 means negative)
+     * next 11 bits -> exponent
+     * last 52 bits -> mantissa
+     *
+     * frexp gives fraction within the range [0.5, 1)
+     * We multiply it by 2 to change the range to [1.0, 2.0)
+     * and reduce exp by 1, since this is the way doubles
+     * are stored in IEEE-754.
      *
-     * Then optional medium (1 byte) or large exponent (2 bytes, lsb first)
+     * Conversion of the fractional part of the new mantissa
+     * to bits is done by repeatedly multiplying it by 2,
+     * checking if it is greater than 1, and turning the
+     * corresponding bit on, if it is.
      *
-     * Then mantissa (0 iff value is 0)
      */
+    uint64_t result = 0;
 
     bool negative = (v < 0.0);
+    if (negative) {
+	v = -v;
+	result |= (uint64_t)1 << 63;
+    }
 
-    if (negative) v = -v;
-
-    int exp = base256ify_double(v);
-
-    string result;
-
-    if (exp <= 6 && exp >= -7) {
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
-		throw Xapian::InternalError("Insane exponent in floating point number");
-	    }
-	    result += negative ? char(0x8f) : char(0x0f);
-	    result += char(unsigned(exp + 32768) & 0xff);
-	    result += char(unsigned(exp + 32768) >> 8);
-	}
+    int exp;
+    v = frexp(v, &exp);
+    v *= 2.0;
+
+    if (exp == 0 && v == 0.0) {
+	result = 0;
+	return string(reinterpret_cast<const char *>(&result),
+		      sizeof(uint64_t));
     }
 
-    int maxbytes = min(MAX_MANTISSA_BYTES, 8);
+    exp += 1022;
 
-    size_t n = result.size();
-    do {
-	unsigned char byte = static_cast<unsigned char>(v);
-	result += char(byte);
-	v -= double(byte);
-	v *= 256.0;
-    } while (v != 0.0 && --maxbytes);
+    result |= (uint64_t)exp << 52;
 
-    n = result.size() - n;
-    if (n > 1) {
-	Assert(n <= 8);
-	result[0] = static_cast<unsigned char>(result[0] | ((n - 1) << 4));
+    // mantissa bit pointer
+    uint64_t mbp = (uint64_t)1 << 51;
+
+    v -= 1.0;
+
+    for (int i = 51; i >= 0; --i) {
+	v *= 2.0;
+	if (v >= 1.0) {
+	    result |= mbp;
+	    v -= 1.0;
+	}
+	mbp >>= 1;
```

Comment:
Working one bit at a time seems rather inefficient.

Can't we at least work byte-by-byte?  Multiply by 256.0, get the byte as the integer truncation of the double, then subtract the byte value from the double?  That's what the existing `serialise_double()` code does.  Here the first byte is only a partial byte (3 bits) but that's not hard to handle.

We could probably work in 4-byte chunks (multiply by 4294967296.0, get the uint32_t as the integer truncation of the double, subtract that from the double.  I think the existing code doesn't do that because it produces a variable length encoding and wants to stop when the remaining double is 0.0, but here we produce a fixed size result.

It might even be possible to just multiply by 2251799813685248.0 and cast to a uint64_t to get all the bits in one go.  We need to make sure the intermediate result is always representable as a double (remembering that double likely doesn't have IEEE properties here).

Reply:
I couldn't really figure out a reason to prefer 32 bits over 26, So I modified it to use 26 bit chunks. Please correct me if I'm wrong.  
Multiplying by 2^51 does not seem like a safe option, so I left it.

Reply:
Changed the function to serialise in one go.